src/System.Net.WebSockets.Client/src/System/Net/WebSockets/WebSocketHandle.Unix.cs

```diff
@@ -2,108 +2,1037 @@
+            {
+                await _sendFrameAsyncLock.WaitAsync().ConfigureAwait(false);
+                try
+                {
+                    EnsureBufferLength(ref _sendBuffer, buffer.Count + MaxMessageHeaderLength);
+
+                    int maskOffset = WriteHeader(opcode, _sendBuffer, buffer, endOfMessage);
+                    int headerLength = maskOffset + MaskLength;
+
+                    if (buffer.Count > 0)
+                    {
+                        for (int i = 0; i < buffer.Count; i++)
+                        {
+                            _sendBuffer[i + headerLength] = (byte)(buffer.Array[buffer.Offset + i] ^ _sendBuffer[maskOffset + (i % MaskLength)]);
+                        }
+                    }
+
+                    await _stream.WriteAsync(_sendBuffer, 0, headerLength + buffer.Count, cancellationToken).ConfigureAwait(false);
+                }
+                catch (ObjectDisposedException ode)
+                {
+                    throw new WebSocketException(WebSocketError.ConnectionClosedPrematurely, ode);
+                }
+                finally
+                {
+                    _sendFrameAsyncLock.Release();
+                }
+            }
+
+            private void SendKeepAliveFrameAsync()
+            {
+                Task pingTask = SendFrameAsync(new ArraySegment<byte>(Array.Empty<byte>()), MessageOpcode.Ping, true, CancellationToken.None);
+                // This is exists purely to keep the connection alive; ignore any failures.
+            }
+
+            private static int WriteHeader(MessageOpcode opcode, byte[] sendBuffer, ArraySegment<byte> payload, bool endOfMessage)
+            {
+                // Client header format:
+                // 1 bit - FIN - 1 if this is the final fragment in the message (it could be the only fragment), otherwise 0
+                // 1 bit - RSV1 - Reserved - 0
+                // 1 bit - RSV2 - Reserved - 0
+                // 1 bit - RSV3 - Reserved - 0
+                // 4 bits - Opcode - How to interpret the payload
+                //     - 0x0 - continuation
+                //     - 0x1 - text
+                //     - 0x2 - binary
+                //     - 0x8 - connection close
+                //     - 0x9 - ping
+                //     - 0xA - pong
+                //     - (0x3 to 0x7, 0xB-0xF - reserved)
+                // 1 bit - Masked - 1 if the payload is masked, 0 if it's not.  Must be 1 for the client
+                // 7 bits, 7+16 bits, or 7+64 bits - Payload length
+                //     - For length 0 through 125, 7 bits storing the length
+                //     - For lengths 126 through 2^16, 7 bits storing the value 126, followed by 16 bits storing the length
+                //     - For lengths 2^16+1 through 2^64, 7 bits storing the value 127, followed by 64 bytes storing the length
+                // 4 bits - Mask - 32-bit random value XOR'd with each 4 bytes of the payload, round-robin
```

Comment:
Reading this as written here this would be misaligned data.  I think you mean 4 BYTES / 32 bits? :smile: